inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/util/QueryUtil.java

```diff
@@ -67,58 +68,146 @@
+     * desired result.
+     *
+     * @param aTypedString typed string from the user
+     * @param aMention the marked surface form
+     * @param aDescriptionIri KB-specific IRI that indicates a description
      * @return a query to retrieve candidate entities
      */
-    public static TupleQuery generateCandidateQuery(RepositoryConnection conn, String tokens,
-        int limit, IRI aDescriptionIri)
+    public static TupleQuery generateCandidateExactQuery(RepositoryConnection conn,
+        String aTypedString, String aMention, IRI aDescriptionIri)
     {
+        aTypedString = RenderUtils.escape(aTypedString);
+        aMention = RenderUtils.escape(aMention);
+
+        // Matching user input exactly
+        String exactMatchingTypedString = getExactMatchingQueryPart("exactTyped");
+
+        // Match surface form exactly
+        String exactMatchingMention = getExactMatchingQueryPart("exactMention");
+
         String query = String.join("\n",
             "DEFINE input:inference 'instances'",
             SPARQL_PREFIX,
-            "SELECT DISTINCT ?e2 ?altLabel ?label ?description WHERE",
+            "SELECT DISTINCT ?e2 ?label ?description WHERE",
             "{",
             "  {",
+            exactMatchingTypedString,
+            "  } ",
+            "  UNION",
+            "  {",
+            exactMatchingMention,
+            "  }",
+            "  FILTER EXISTS { ?e2 ?p ?v }",
+            "  FILTER NOT EXISTS ",
+            "  {",
+            "    VALUES ?topic {" + String.join(" ", WIKIMEDIA_INTERNAL,
+                WIKIMEDIA_PROJECT_PAGE, WIKIMEDIA_CATEGORY, WIKIMEDIA_DISAMBIGUATION_PAGE,
+                WIKIMEDIA_LIST_ARTICLE, WIKIMEDIA_TEMPLATE, WIKIMEDIA_NEWS_ARTICLE,
+                WIKIMEDIA_NAVIGATIONAL_TEMPLATE) +
+                "}",
+            "    ?e2 rdf:type ?topic",
+            "  }",
+            "  ?e2 rdfs:label ?label.",
+            "  FILTER ( lang(?label) = \"en\" )",
+            "}");
+
+        ValueFactory vf = SimpleValueFactory.getInstance();
+
+        TupleQuery tupleQuery = conn.prepareTupleQuery(QueryLanguage.SPARQL, query);
+        tupleQuery.setBinding("exactTyped", vf.createLiteral(aTypedString));
+        tupleQuery.setBinding("exactMention", vf.createLiteral(aMention));
+
+        tupleQuery.setBinding("descriptionIri", aDescriptionIri);
+        return tupleQuery;
+    }
+
+    private static String getFullTextMatchingQueryPart(String aString, int aLimit) {
+        return  String.join("\n",
+            "    SELECT DISTINCT ?e2 ?altLabel ?description WHERE",
             "    {",
             "      VALUES ?labelpredicate {rdfs:label skos:altLabel}",
-            "      GRAPH " + TERMS,
             "      {",
             "        ?e2 ?labelpredicate ?altLabel.",
-            "        ?altLabel bif:contains '?entityLabel'. ",
+            "        ?altLabel bif:contains '?" + aString + "'. ",
```

Comment:
This opens the way to query injection problems. If the query string includes a double quote, the SPARQL query breaks. I see that some escaping is done in other locations - but it would be better to do the escaping where it is needed, namely when constructing the query.